### DISCUSSION TITLE
Create3 teleop example call from python script

### DISCUSSION THREAD

**Discussion by azbitak-invn** [iRobotEducation/create3_docs#310] [Questions]
I'm working with Irobot development kit and we originally had python scripts to drive pycreate 2 (connected to a raspberry pi, python scripts using pycreate2 python lib).
We wanted to move to pycreate3, so I followed the tutorial to install webserver on raspberry pi 4 using ROS2. And I m now able to play with the robot .

I m interested on the example teleop in order to know how to call the robot messages into python scripts.
My first guess was to "connect" my legacy script in order to replace motion command by the one from teleop. Unfortunately I can't find how to call basic robot motion. In the python_example folder I can't find any api, is there using a library that I can reuse for my scripts?
What do you suggest ?

Many thanks

**Comment by shamlian** [comment]
Hi there,

I'd like to help, but I am not sure I completely understand the problem. It sounds like you were hoping that the Create 3 robot would be compatible with the Create 1 and Create 2 OI, and now are working to port Python scripts to work with Create 3. There are two ways you could do this, depending on what you are doing with these robots. One way you could do it is to use the BETA [iRobot Education Python SDK](https://github.com/iRobotEducation/irobot-edu-python-sdk), which seeks to emulate level 3 of [iRobot Coding](code.irobot.com), and connects via a Bluetooth connection. If you are doing things at a relatively low frequency, this might work for you. If not, you should connect over either Wi-Fi or Ethernet over USB, and use a Python ROS 2 client library like `rospy` to control your robot. What `python_example` folder are you referring to? If you give me an idea of the interface you're looking for, we might be able to help you better. Let us know!

**Reply by azbitak-invn** [reply]
Hello,

Yes the idea was to reuse my original scripts to control the robot with python (previously we were using this api [https://pypi.org/project/pycreate2/](url) Didn't know there was a BETA with SDK python, I ll look at it, but as it s a beta the API might change a lot until release stage.
My first guess was to re-use this kind of example (here turn action)  [https://github.com/paccionesawyer/Create3_ROS2_Intro/blob/main/individual_examples/action_turn.py](url) and instead of writing a main, adding a function turn. Then call it from my python script, but this approach doesn't seems to work as I expect.
Could you advice me ?
Thanks,

**Reply by shamlian** [reply]
If you're looking to learn ROS 2 with the Create 3 robot, another user was able to use the tutorials from Tufts University linked to in this question: https://github.com/iRobotEducation/create3_docs/discussions/202

**Reply by shamlian** [reply]
If you provide specific code and runtime messages that don't work, we can try to help you debug that. Be sure to also include the version of your robot, the version of ROS 2 running on your computer, and which middleware you've chosen.

**Reply by azbitak-invn** [reply]
Thanks, appreciate your help. The idea is more to reuse codes that you linked https://github.com/iRobotEducation/create3_docs/discussions/202 (ex telerobot s example) and instead of triggering robot motion commands from keyboard, triggered them from my python scripts. But could not find clear api to call

**Reply by azbitak-invn** [reply]
I think safest is using ros2 as it s more robust as BETA sdk for now. So I used a set of basic class node to have a poc python script using : 
* dock
* drive
* undock
* rotate
[test-robovac.txt](https://github.com/iRobotEducation/create3_docs/files/10444314/test-robovac.txt)
(attached .txt which is test-robovac.py)
The script is very simple and all functions are flat, it s just to know if it's the right way to proceed. Each class are called from functions which are setting some parameters and then I called them sequentially from a main.

I have two question : 
* It seems the robot is not doing the drive, I got no error but the robot is not moving : 
`
drive
[INFO] [1674034968.733365694] [drive_distance_action_client]: Goal accepted :)
[INFO] [1674034968.745889071] [drive_distance_action_client]: Result: irobot_create_msgs.action.DriveDistance_Result(pose=geometry_msgs.msg.PoseStamped(header=std_msgs.msg.Header(stamp=builtin_interfaces.msg.Time(sec=1674034257, nanosec=565949655), frame_id='odom'), pose=geometry_msgs.msg.Pose(position=geometry_msgs.msg.Point(x=-0.9712210297584534, y=-0.20120835304260254, z=-0.02262713387608528), orientation=geometry_msgs.msg.Quaternion(x=-0.006014670258410334, y=-0.000869048761621458, z=0.8998564159505865, w=0.436143897318875))))
`
Is it because its relative position vs absolute?

* How could I had condition with bulmper sensor on this? For example if bumpet are set, I ll do a 180 degree rotation.

Thank you!

**Reply by kwujciak** [reply]
Oftentimes when I get the info message but no error, that is an issue with my wifi connection. I found that when the connection was stable and strong, the robot would successfully carry out the action (along with the info message).

**Reply by azbitak-invn** [reply]
In parallel I folowed the way "to emulate level 3 of [iRobot Coding](https://github.com/iRobotEducation/create3_docs/discussions/code.irobot.com), and connects via a Bluetooth connection." But seems I can't connect bluetooth with raspberryPi. I got 
`    raise BleakDeviceNotFoundError(
bleak.exc.BleakDeviceNotFoundError: Device with address 00:16:A4:64:DF:7F was not found. It may have been removed from BlueZ when scanning stopped.
`
Someone already got this ?

**Reply by shamlian** [reply]
I suspect you may be encountering https://github.com/iRobotEducation/irobot-edu-python-sdk/issues/3 .

**Reply by azbitak-invn** [reply]
Thanks, It seems that when calling python script it try to connect to the proper robot (the name is the one I set from webserver).
' python3 bumper_control.py
Connecting to Create3-gre (00:16:A4:64:DF:7F)
'
But it then fail with the same message.
'bluetoothctl devices' shows nothing and 'bluetoothctl remove 00:16:A4:64:DF:7F' give "Device 00:16:A4:64:DF:7F not available"

So I m not sure it s supported by the RPI natively. It s BLE or normal Bluetooth ?

I also have an other question concerning the embedded server using ROS2 with robot. It seems required to have the usb connected to the RPI AND the both connected to the same wifi. Is there a reason of that ? Why both should be connected (wifi and usb) ?

**Reply by shamlian** [reply]
iRobot Education's robots communicate using a common Bluetooth Low Energy serial protocol, the details of which can be found [here](https://github.com/iRobotEducation/root-robot-ble-protocol). The web browser uses a `web-bluetooth` backend while the native OS uses a Python cross-platform backend called [`bleak`](https://github.com/hbldh/bleak). Otherwise, they are nearly the same implementations of [iRobot Coding Level 3](code.robot.com). If your OS and radio support BLE, it should work. To be sure -- what version of the downloadable Python SDK are you running (how did you install it)?

As far as using ROS 2 with the robot: sounds like you might have a problem with your middleware. Are you using CycloneDDS or Fast-DDS? See [this](https://iroboteducation.github.io/create3_docs/setup/xml-config/)

**Reply by azbitak-invn** [reply]
Thanks, appreciate your support.
Still trying to know which approach is the best.

For the python SDK I installed it by following installation using installation from sources https://github.com/iRobotEducation/irobot-edu-python-sdk . I m able to connect the robot with bluetooth when running the shell example from https://github.com/iRobotEducation/root-robot-ble-protocol/blob/main/examples/robot-driver.sh . With this script the robot is connecting properly but with the python script I got this '[00:16:a4:64:df:7f] Discovered: Create3-gre
Press Enter to drive robot, CTRL-C to quit
[00:16:a4:64:df:7f] Discovered: None

Traceback (most recent call last):
  File "/home/pi/root-robot-ble-protocol/examples/drive-root.py", line 68, in <module>
    manager.robot.drive_forward()
  File "/home/pi/root-robot-ble-protocol/examples/drive-root.py", line 47, in drive_forward
    self.tx_characteristic.write_value([0x01, 0x04, 0x00, 0x00, 0x00, 0x00,
'
With ROS2 I followed this tutorial (using cycoleDDS) https://jimbobbennett.dev/blogs/irobot-create3-connect-a-pi/ at the end it works (can communicate with robot and sending goals) but had to connect both wifi and usb cable which I find strange.

Regards

**Reply by shamlian** [reply]
What OS and version are you running on your Raspberry Pi 4? Also, can you please tell me what the commit hash is that you're using (`git rev-parse HEAD`) in case there is a bug?

**Reply by azbitak-invn** [reply]
I use 2 images, 1 for ROS2 and one for the bluetooth.
I think you ask for the bluetooth case ? Othewise for ROS2 I m using ubuntu server (the one from the tutorial).
For the "simple" bluetooth usage, I m using : 
'Linux raspberrypi 5.15.84-v7l+ #1613 SMP Thu Jan 5 12:01:26 GMT 2023 armv7l GNU/Linux'

git tag for root-robot-ble-protocol/ is d3e049fb96c78f6931d128d86b801d23a15eae26

git tag for Tufts_Create3_Examples is 56ccd7ced3f20212f95461f81a7288135edfa792

And sorry in fact I used pip3 install irobot_edu_sdk

**Reply by azbitak-invn** [reply]
Hi @shamlian,

I did some advancement on the bluetooth connection with raspberryPi and create3.
I m able to connect to the robot using "bluetoothctl" but I have to do a workaround otherwise I suspect getting the same failure as using the python sdk ( 'raise BleakDeviceNotFoundError( bleak.exc.BleakDeviceNotFoundError: Device with address 00:16:A4:64:DF:7F was not found. It may have been removed from BlueZ when scanning stopped.')

The workaround is I havce to proceed this step to have a connection with the robot : 
`bluetoothctl
menu scan
uuids 48c5d828-ac2a-442d-97a3-0c9822b04979
back 
scan on
connect  00:16:A4:64:DF:7F
scan off`

If I do the connect after scan off, the device is removed. I suspect there is the same issue on SDK with BlueZ. But couldn't find where I could fix it. If you have an idea ?

Thanks
